docs: Add documentation for avoiding hardcoding in tests (#16076)
hashicorp/terraform-provider-aws: docs/contributing/running-and-writing-acceptance-tests.md

- [Acceptance Test Sweepers](#acceptance-test-sweepers)
    - [Running Test Sweepers](#running-test-sweepers)
    - [Writing Test Sweepers](#writing-test-sweepers)
- [Acceptance Test Checklists](#acceptance-test-checklists)
    - [Basic Acceptance Test Design](#basic-acceptance-test-design)
    - [Test Implementation](#test-implementation)
    - [Avoid Hard Coding](#avoid-hard-coding)
        - [Hardcoded Account IDs](#hardcoded-account-ids)
        - [Hardcoded AMI IDs](#hardcoded-ami-ids)
        - [Hardcoded Availability Zones](#hardcoded-availability-zones)
        - [Hardcoded Database Versions](#hardcoded-database-versions)
        - [Hardcoded Instance Types](#hardcoded-instance-types)
        - [Hardcoded Partition DNS Suffix](#hardcoded-partition-dns-suffix)
        - [Hardcoded Partition in ARN](#hardcoded-partition-in-arn)
        - [Hardcoded Region](#hardcoded-region)
        - [Hardcoded Spot Price](#hardcoded-spot-price)

>         - [Hardcoded State Hash](#hardcoded-state-hash)

Reviewer: ```suggestion
```

}

resource "aws_db_instance" "bar" {
  engine               = data.aws_rds_engine_version.default.engine
  engine_version       = data.aws_rds_engine_version.default.version
  instance_class       = data.aws_rds_orderable_db_instance.test.instance_class
  skip_final_snapshot  = true
  parameter_group_name = "default.${data.aws_rds_engine_version.default.parameter_group_family}"
}
```

#### Hardcoded Instance Types

- [ ] __Uses Instance Type Data Source__: Singular hardcoded instance types and classes, e.g., `t2.micro` and `db.t2.micro`, should be replaced with a list of preferences using a data source. Because offerings vary from region to region and partition to partition, providing a list of preferences dramatically improves the likelihood that one of the options will be available. Depending on the situation, there are several data sources for instance types and classes, including:
    - [`aws_ec2_instance_type_offering` data source](https://www.terraform.io/docs/providers/aws/d/ec2_instance_type_offering.html) - Convenience functions declare configurations that are referenced with `data.aws_ec2_instance_type_offering.available` including:

>         - The `testAccAvailableEc2InstanceTypeForAvailabilityZone()` function (defined in `resource_aws_instance_test.go`) for test configurations using an EC2 Subnet which is inherently within a single Availability Zone
>         - The `testAccAvailableEc2InstanceTypeForRegion()` function (defined in `resource_aws_instance_test.go`) for test configurations that do not include specific Availability Zones

Reviewer: We should avoid specifying file locations which will change over time and do not affect the underlying code discoverability

```suggestion
        - The `testAccAvailableEc2InstanceTypeForAvailabilityZone()` function for test configurations using an EC2 Subnet which is inherently within a single Availability Zone
        - The `testAccAvailableEc2InstanceTypeForRegion()` function for test configurations that do not include specific Availability Zones
```

Here's an example of using `aws_partition` and `data.aws_partition.current.dns_suffix`:

```hcl
data "aws_partition" "current" {}

resource "aws_iam_role" "test" {
  assume_role_policy = <<POLICY
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "Service": "cloudtrail.${data.aws_partition.current.dns_suffix}"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
POLICY
}
```

#### Hardcoded Partition in ARN

- [ ] __Uses aws_partition Data Source__: Any hardcoded AWS Partition configuration, e.g. the `aws` in a `arn:aws:SERVICE:REGION:ACCOUNT:RESOURCE` ARN, should be replaced with the [`aws_partition` data source](https://www.terraform.io/docs/providers/aws/d/partition.html). A common pattern is declaring `data "aws_partition" "current" {}` and referencing it via `data.aws_partition.current.partition`.

> - [ ] __Uses Builtin ARN Check Functions__: Tests should utilize available ARN check functions, e.g. `testAccMatchResourceAttrRegionalARN()`, to validate ARN attribute values in the Terraform state over `resource.TestCheckResourceAttrSet()` and `resource.TestMatchResourceAttr()`.

Reviewer: While we are mentioning this one function, we should probably include the other common ones:

```suggestion
- [ ] __Uses Builtin ARN Check Functions__: Tests should utilize available ARN check functions to validate ARN attribute values in the Terraform state over `resource.TestCheckResourceAttrSet()` and `resource.TestMatchResourceAttr()`:

    - `testAccCheckResourceAttrRegionalARN()` verifies an ARN matches the account ID and region of the test execution and an exact resource value
    - `testAccMatchResourceAttrRegionalARN()` verifies an ARN matches the account ID and region of the test execution and a regular expression of the resource value
    - `testAccCheckResourceAttrGlobalARN()` verifies an ARN matches the account ID of the test execution and an exact resource value
    - `testAccMatchResourceAttrGlobalARN()` verifies an ARN matches the account ID of the test execution and a regular expression of the resource value
```

    - [`aws_canonical_user_id` data source](https://www.terraform.io/docs/providers/aws/d/canonical_user_id.html),
    - [`aws_billing_service_account` data source](https://www.terraform.io/docs/providers/aws/d/billing_service_account.html), and
    - [`aws_sagemaker_prebuilt_ecr_image` data source](https://www.terraform.io/docs/providers/aws/d/sagemaker_prebuilt_ecr_image.html).
- [ ] __Uses testAccCheckResourceAttrAccountID()__: Tests should utilize the available AWS Account ID check function, `testAccCheckResourceAttrAccountID()` to validate account ID attribute values in the Terraform state over `resource.TestCheckResourceAttrSet()` and `resource.TestMatchResourceAttr()`

Here's an example of using `aws_caller_identity`:

```hcl
data "aws_caller_identity" "current" {}

resource "aws_backup_selection" "test" {
  plan_id      = aws_backup_plan.test.id
  name         = "tf_acc_test_backup_selection_%[1]d"
  iam_role_arn = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/service-role/AWSBackupDefaultServiceRole"
}
```

#### Hardcoded AMI IDs

> - [ ] __Uses aws_ami Data Source__: Any hardcoded AMI ID configuration, e.g. `ami-12345678`, should be replaced with the [`aws_ami` data source](https://www.terraform.io/docs/providers/aws/d/ami.html) pointing to an Amazon Linux image. Use the convenience function called `testAccLatestAmazonLinuxHvmEbsAmiConfig()` (defined in `resource_aws_instance_test.go`) to declare `data "aws_ami" "amzn-ami-minimal-hvm-ebs" {...}`. Other convenience functions are also available including `testAccLatestAmazonLinuxHvmInstanceStoreAmiConfig()`, `testAccLatestAmazonLinuxPvEbsAmiConfig()`, `testAccLatestAmazonLinuxPvInstanceStoreAmiConfig`, and `testAccLatestWindowsServer2016CoreAmiConfig()`.

Reviewer: I would recommend listing out the configuration helper functions and not calling out the physical location of functions which will change over time:

```suggestion
- [ ] __Uses aws_ami Data Source__: Any hardcoded AMI ID configuration, e.g. `ami-12345678`, should be replaced with the [`aws_ami` data source](https://www.terraform.io/docs/providers/aws/d/ami.html) pointing to an Amazon Linux image. The codebase includes test configuration helper functions to simplify these lookups:
    - `testAccLatestAmazonLinuxHvmEbsAmiConfig()`: The most recommended AMI lookup using Amazon Linux, HVM virtualization, and EBS storage. To reference the AMI ID in the test configuration: `data.aws_ami.amzn-ami-minimal-hvm-ebs.id`.
    - `testAccLatestAmazonLinuxHvmInstanceStoreAmiConfig()`: AMI lookup using Amazon Linux, HVM virtualization, and Instance Store storage. Should only be used in testing that requires that type of storage over EBS. To reference the AMI ID in the test configuration: `data.aws_ami.amzn-ami-minimal-hvm-instance-store.id`.
    - `testAccLatestAmazonLinuxPvEbsAmiConfig()`: AMI lookup using Amazon Linux, Paravirtual virtualization, and EBS storage. Should only be used in testing that requires Paravirtual virtualization over HVM. To reference the AMI ID in the test configuration: `data.aws_ami.amzn-ami-minimal-pv-ebs.id`.
    - `testAccLatestAmazonLinuxPvInstanceStoreAmiConfig`: AMI lookup using Amazon Linux, Paravirtual virtualization, and Instance Store storage. Should only be used in testing that requires Paravirtual virtualization over HVM and Instance Store over EBS. To reference the AMI ID in the test configuration: `data.aws_ami.amzn-ami-minimal-pv-instance-store.id`.
    - `testAccLatestWindowsServer2016CoreAmiConfig()`: AMI lookup using Windows Server 2016 Core, HVM virtualization, and EBS storage. Should only be used in testing that requires Windows. To reference the AMI ID in the test configuration: `data.aws_ami.win2016core-ami.id`.
```

  instance_type = "m1.small"
}
`, rName))
}
```

#### Hardcoded Availability Zones

- [ ] __Uses aws_availability_zones Data Source__: Any hardcoded AWS Availability Zone configuration, e.g. `us-west-2a`, should be replaced with the [`aws_availability_zones` data source](https://www.terraform.io/docs/providers/aws/d/availability_zones.html). Use the convenience function called `testAccAvailableAZsNoOptInConfig()` (defined in `resource_aws_instance_test.go`) to declare `data "aws_availability_zones" "available" {...}`. You can then reference the data source via `data.aws_availability_zones.available.names[0]` or `data.aws_availability_zones.available.names[count.index]` in resources utilizing `count`.

Here's an example of using `testAccAvailableAZsNoOptInConfig()` and `data.aws_availability_zones.available.names[0]`:

```go
func testAccAwsInstanceVpcConfigBasic(rName string) string {
	return composeConfig(testAccAvailableAZsNoOptInConfig(), fmt.Sprintf(`
resource "aws_subnet" "test" {
  availability_zone = data.aws_availability_zones.available.names[0]
  cidr_block        = "10.0.0.0/24"
  vpc_id            = aws_vpc.test.id
}
`, rName))
}
```

> #### Hardcoded Database Versions

Reviewer: This will likely include other types of versions in the future.

```suggestion
#### Hardcoded Versions
```

> - [ ] __Uses Version Data Sources__: Singular hardcoded database versions, e.g., `5.7.42`, should be removed (which means the AWS-defined default version will be used) or replaced with a list of preferred versions using a data source. Because versions change over times and version offerings vary from region to region and partition to partition, using the default version or providing a list of preferences ensures a version will be available. Depending on the situation, there are several data sources for versions, including:

Reviewer: ```suggestion
- [ ] __Uses Version Data Sources__: Hardcoded versions, e.g. RDS MySQL Engine Version `5.7.42`, should be removed (which means the AWS-defined default version will be used) or replaced with a list of preferred versions using a data source. Because versions change over times and version offerings vary from region to region and partition to partition, using the default version or providing a list of preferences ensures a version will be available. Depending on the situation, there are several data sources for versions, including:
```

    - [`aws_ec2_instance_type_offering` data source](https://www.terraform.io/docs/providers/aws/d/ec2_instance_type_offering.html) (use the convenience function called `testAccAvailableEc2InstanceTypeForRegion()` or `testAccAvailableEc2InstanceTypeForAvailabilityZone()` (defined in `resource_aws_instance_test.go`) to declare `data "aws_ec2_instance_type_offering" "available" {...}`),
    - [`aws_rds_orderable_db_instance` data source](https://www.terraform.io/docs/providers/aws/d/rds_orderable_db_instance.html),
    - [`aws_neptune_orderable_db_instance` data source](https://www.terraform.io/docs/providers/aws/d/neptune_orderable_db_instance.html), and
    - [`aws_docdb_orderable_db_instance` data source](https://www.terraform.io/docs/providers/aws/d/docdb_orderable_db_instance.html).

> 

Reviewer: ```suggestion

The codebase provides test configuration helper functions to simplify these lookups including:

    - `testAccAvailableEc2InstanceTypeForAvailabilityZone()`: Lookup available EC2 Instance Types for an Availability Zone within a Region. This is recommended for any configuration using an EC2 Subnet, which inherently is within a single Availability Zone. To reference the instance type in the test configuration: `data.aws_ec2_instance_type_offering.available.instance_type`.
    - `testAccAvailableEc2InstanceTypeForRegion()`: Lookup available EC2 Instance Types for a Region. This should only be used when the test configuration does not include specific Availability Zones. To reference the instance type in the test configuration: `data.aws_ec2_instance_type_offering.available.instance_type`.

```

```go
func testAccAWSSpotInstanceRequestConfig(rInt int) string {
	return composeConfig(testAccAvailableEc2InstanceTypeForRegion("t3.micro", "t2.micro"),
		fmt.Sprintf(`
resource "aws_spot_instance_request" "test" {
  instance_type        = data.aws_ec2_instance_type_offering.available.instance_type
  spot_price           = "0.05"
  wait_for_fulfillment = true
}
`, rInt))
}
```

Here's an example of using `aws_rds_orderable_db_instance` and `data.aws_rds_orderable_db_instance.test.instance_class`:

```hcl
data "aws_rds_orderable_db_instance" "test" {
  engine                     = "mysql"
  engine_version             = "5.7.31"
  preferred_instance_classes = ["db.t3.micro", "db.t2.micro", "db.t3.small"]
}

resource "aws_db_instance" "test" {
  engine              = data.aws_rds_orderable_db_instance.test.engine
  engine_version      = data.aws_rds_orderable_db_instance.test.engine_version
  instance_class      = data.aws_rds_orderable_db_instance.test.instance_class
  skip_final_snapshot = true
  username            = "test"
}
```

#### Hardcoded Partition DNS Suffix

> - [ ] __Uses aws_partition Data Source__: Any hardcoded DNS suffix configuration, e.g. the `amazonaws.com` in a `ec2.amazonaws.com` service ID, should be replaced with the [`aws_partition` data source](https://www.terraform.io/docs/providers/aws/d/partition.html). A common pattern is declaring `data "aws_partition" "current" {}` and referencing it via `data.aws_partition.current.dns_suffix`.

Reviewer: Nit: "principal"

```suggestion
- [ ] __Uses aws_partition Data Source__: Any hardcoded DNS suffix configuration, e.g. the `amazonaws.com` in a `ec2.amazonaws.com` service principal, should be replaced with the [`aws_partition` data source](https://www.terraform.io/docs/providers/aws/d/partition.html). A common pattern is declaring `data "aws_partition" "current" {}` and referencing it via `data.aws_partition.current.dns_suffix`.
```

Here's an example of using `aws_region` and `data.aws_region.current.name`:

```hcl
data "aws_region" "current" {}

resource "aws_route53_zone" "test" {
  vpc {
    vpc_id     = aws_vpc.test.id
    vpc_region = data.aws_region.current.name
  }
}
```

#### Hardcoded Spot Price

- [ ] __Uses aws_ec2_spot_price Data Source__: Any hardcoded spot prices, e.g. `0.05`, should be replaced with the [`aws_ec2_spot_price` data source](https://www.terraform.io/docs/providers/aws/d/ec2_spot_price.html). A common pattern is declaring `data "aws_ec2_spot_price" "current" {}` and referencing it via `data.aws_ec2_spot_price.current.spot_price`.

Here's an example of using `aws_ec2_spot_price` and `data.aws_ec2_spot_price.current.spot_price`:

```hcl
data "aws_ec2_spot_price" "current" {
  instance_type = "t3.medium"

  filter {
    name   = "product-description"
    values = ["Linux/UNIX"]
  }
}

resource "aws_spot_fleet_request" "test" {
  spot_price      = data.aws_ec2_spot_price.current.spot_price
  target_capacity = 2
}
```

> #### Hardcoded State Hash
> 
> - [ ] __Uses tfawsresource.TestCheckTypeSetElemNestedAttrs()__: Any hardcoded state hashes used in test steps, e.g., `resource.TestCheckResourceAttr(resourceName, "port_range.4120185872.from_port", "80")`, should be replaced with `tfawsresource.TestCheckTypeSetElemNestedAttrs()`.
> 
> Here's an example of using `TestCheckTypeSetElemNestedAttrs()`:
> 
> ```go
> func TestAccAwsGlobalAcceleratorListener_basic(t *testing.T) {
> 	resourceName := "aws_globalaccelerator_listener.example"
> 	rInt := acctest.RandInt()
> 
> 	resource.ParallelTest(t, resource.TestCase{
> 		PreCheck:     func() { testAccPreCheck(t); testAccPreCheckGlobalAccelerator(t) },
> 		Providers:    testAccProviders,
> 		CheckDestroy: testAccCheckGlobalAcceleratorListenerDestroy,
> 		Steps: []resource.TestStep{
> 			{
> 				Config: testAccGlobalAcceleratorListener_basic(rInt),
> 				Check: resource.ComposeTestCheckFunc(
> 					testAccCheckGlobalAcceleratorListenerExists(resourceName),
> 					resource.TestCheckResourceAttr(resourceName, "client_affinity", "NONE"),
> 					tfawsresource.TestCheckTypeSetElemNestedAttrs(resourceName, "port_range.*", map[string]string{
> 						"from_port": "80",
> 						"to_port":   "81",
> 					}),
> 				),
> 			},
> 		},
> 	})
> }
> ```

Reviewer: It might be best to leave this section out of our documentation here since these helpers have now been migrated into the Terraform Plugin SDK and ensure the "Extending Terraform" documentation includes how to use them.

```suggestion
```
